Python/optimizer_bytecodes.c

```diff
@@ -412,6 +413,7 @@ dummy_func(void) {
     }
 
     op(_UNARY_INVERT, (value -- res)) {
+        REPLACE_OPCODE_IF_EVALUATES_PURE(value);
```

Comment:
I think this isn't pure for bool. `~True` raises a warning.

Reply:
Does it make sense to guard against booleans? Or are we saying that we can't do that/it's not worth it?

Reply:
We could guard against booleans I think.

Anything that raises a warning isn't pure as that means the optimizer might cause a warning to be emitted during optimizing code, and we don't want that.

To test it out, try typing ~True into the current main's CPython repl. It's a recent deprecation so it wont show up on your system cpython probably, but a warning is emitted.

I found this out because I vaguely recalled reading about it somewhere https://discuss.python.org/t/bool-deprecation/62232